File: Leetcode September Daily Challenge/Leetcode #1457/C++/1457.cpp
#define tn TreeNode 
class Solution {
public:
int ans =0;
vector<int> vis;
void solve(tn *root){
    if(!root)return;
    vis[root->val]++;
    if(root->left==NULL and root->right==NULL){            
        int odd=-1;
        bool ok =true;
        for(int i=1;i<=9;i++){
            if(vis[i] %2   and (odd==-1)){
                odd=1;                    
            }
            else if(vis[i]%2){
                ok=false;
                break;
            }
        }
        if(ok) ans++;            
        
    }        
    solve(root->left);
    solve(root->right);
    vis[root->val]--;
}
int pseudoPalindromicPaths (TreeNode* root) {
    vis.resize(10,0);
    solve(root);
    return ans;
}

File: Leetcode September Daily Challenge/Leetcode #1457/Python/1457.py
//Traverse all paths using dfs() function.
//Track count of digits in the path in counter list.
//Once a leaf is detected check if path is a pseudo-palindromic one. It is true only if counter has zero or one odd element(s).
//Time complexity = O(n), where n is a number of nodes in the tree
//Space complexity = O(h), where h is a height of the tree
class Solution:
    def pseudoPalindromicPaths (self, root: Optional[TreeNode]) -> int:
        def dfs(node: Optional[TreeNode], counter: list[int]) -> int:
            counter[node.val-1] = 0 if counter[node.val-1] else 1
            if not (node.left or node.right):   # it is a leaf
                ret = 0 if sum(counter) > 1 else 1
            else:
                ret = 0
                if node.left:
                    ret = dfs(node.left, counter)
                if node.right:
                    ret += dfs(node.right, counter)
            counter[node.val-1] = 0 if counter[node.val-1] else 1
            return ret
                
        return dfs(root, [0] * 9)
